pkg/asset/machines/aws/master.go

```diff
@@ -2,72 +2,106 @@
-    labels:
-      sigs.k8s.io/cluster-api-cluster: {{$c.ClusterName}}
-      sigs.k8s.io/cluster-api-machine-role: master
-      sigs.k8s.io/cluster-api-machine-type: master
-  spec:
-    providerConfig:
-      value:
-        apiVersion: aws.cluster.k8s.io/v1alpha1
-        kind: AWSMachineProviderConfig
-        ami:
-          id: {{$c.AMIID}}
-        instanceType: {{$c.Machine.InstanceType}}
-        placement:
-          region: {{$c.Region}}
-          availabilityZone: {{$instance.AvailabilityZone}}
-        subnet:
-          filters:
-          - name: "tag:Name"
-            values:
-            - "{{$c.ClusterName}}-master-{{$instance.AvailabilityZone}}"
-        iamInstanceProfile:
-          id: "{{$c.ClusterName}}-master-profile"
-        tags:
-{{- range $key,$value := $c.Tags}}
-          - name: "{{$key}}"
-            value: "{{$value}}"
-{{- end}}
-        securityGroups:
-          - filters:
-            - name: "tag:Name"
-              values:
-              - "{{$c.ClusterName}}_master_sg"
-        userDataSecret:
-          name: "master-user-data"
-    versions:
-      kubelet: ""
-      controlPlane: ""
-{{- end}}
-`))
+		provider := awsprovider.AWSMachineProviderConfig{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: "aws.cluster.k8s.io/v1alpha1",
+				Kind:       "AWSMachineProviderConfig",
+			},
+			InstanceType:       mpool.InstanceType,
+			AMI:                awsprovider.AWSResourceReference{ID: &mpool.AMIID},
+			Tags:               ptags,
+			IAMInstanceProfile: &awsprovider.AWSResourceReference{ID: func(s string) *string { return &s }(fmt.Sprintf("%s-master-profile", clustername))},
+			UserDataSecret:     &corev1.LocalObjectReference{Name: userDataSecret},
+			Subnet: awsprovider.AWSResourceReference{
+				Filters: []awsprovider.Filter{{
+					Name:   "tag:Name",
+					Values: []string{fmt.Sprintf("%s-master-%s", clustername, az)},
+				}},
+			},
+			Placement: awsprovider.Placement{Region: platform.Region, AvailabilityZone: az},
+			SecurityGroups: []awsprovider.AWSResourceReference{{
+				Filters: []awsprovider.Filter{{
+					Name:   "tag:Name",
+					Values: []string{fmt.Sprintf("%s_master_sg", clustername)},
+				}},
+			}},
+		}
+
+		name := fmt.Sprintf("%s-%s-%d", clustername, pool.Name, idx)
+		machine := clusterapi.Machine{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: "cluster.k8s.io/v1alpha1",
+				Kind:       "Machine",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: "openshift-cluster-api",
+				Name:      name,
+				Labels: map[string]string{
+					"sigs.k8s.io/cluster-api-cluster":      clustername,
+					"sigs.k8s.io/cluster-api-machine-role": "master",
+					"sigs.k8s.io/cluster-api-machine-type": "master",
+				},
+			},
+			Spec: clusterapi.MachineSpec{
+				ProviderConfig: clusterapi.ProviderConfig{
+					Value: &runtime.RawExtension{Object: &provider},
+				},
+				Versions: clusterapi.MachineVersionInfo{
+					Kubelet:      "",
+					ControlPlane: "",
```

Comment:
I'd [still][1] rather leave `Versions` unset and file issues or set a FIXME about these two parameters.  Setting them to empty strings doesn't make sense to me.

[1]: https://github.com/openshift/installer/pull/567#discussion_r229449615

Reply:
This is explicitly saying don't care about version for these we have diff mechanism for controlling them.

Reply:
> This is explicitly saying don't care about version for these we have diff mechanism for controlling them.

But you can use zero-value for that, no?

```console
$ cat test.go 
package main

import (
	"fmt"
	clusterapi "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
)

func main() {
	spec := clusterapi.MachineSpec{}
	fmt.Printf("%q %t\n", spec.Versions.Kubelet, spec.Versions.Kubelet == "")
}
$ go run test.go 
"" true
```